### PR TITLE
Store Date as UTC and show it in server time zone

### PR DESCRIPTION
**Pull request by astridx**
Pull Request for Issue https://github.com/joomla-projects/gsoc18_override_management/pull/52#issuecomment-408002797.

### Summary of Changes

The values for modified and created date are stored in UTC, but shown in server time zone.
If your server time zone is Australia 
global configuration test administration
you see this time:
templates customise cassiopeia test administration 4
If your server time zone is in Europe 
global configuration test administration 1
you see this time:
templates customise cassiopeia test administration 5

For Europe you see in the view this values:
templates customise cassiopeia test administration 6
In the database the UTC values are stored
localhost localhost joomla_db a_template_overrides phpmyadmin 4 5 4 1deb2ubuntu2

### CODE REVIEW

File: administrator/components/com_templates/Model/TemplateModel.php [hunk 1]
```diff
@@ -209,6 +209,26 @@ public function getUpdatedList($state = false, $all = false, $cleanup = false)
...
 		}
 
+		// Get timezone for converting the created and modified date.
+		$tz = new \DateTimeZone(Factory::getApplication()->get('offset'));
+
+		foreach ($results as $result)
+		{
+			if ((int) $result->created_date)
+			{
+				$created_date = new Date($result->created_date);
```
Comment (laoneo) [inline administrator/components/com_templates/Model/TemplateModel.php]: In what for a timezone is this stored?

File: plugins/installer/override/override.php [hunk 2]
```diff
@@ -336,10 +336,8 @@ private function saveOverrides($pks)
 		{
 			$insertQuery->clear('values');
 
-			$tz = new \DateTimeZone(Factory::getApplication()->get('offset'));
 			$date = new Date('now');
```
Comment (astridx) [inline plugins/installer/override/override.php]: @laoneo If I am right the timezone of `$date = new Date('now');`
is UTC as this is the default time zone.

Comment (laoneo) [reply plugins/installer/override/override.php]: I guess so too.

**Comment by Anu1601CS** [comment]
@astridx Can we do this without creating a new loop. Mean, here

https://github.com/joomla-projects/gsoc18_override_management/blob/2c393ffef6c4f6dcab6c6beb9aa60ffe9b2a5eaa/administrator/components/com_templates/Model/TemplateModel.php#L202

File: administrator/components/com_templates/Model/TemplateModel.php [hunk 3]
```diff
@@ -209,6 +209,26 @@ public function getUpdatedList($state = false, $all = false, $cleanup = false)
 			}
 		}
 
+		// Get timezone for converting the created and modified date.
+		$tz = new \DateTimeZone(Factory::getApplication()->get('offset'));
+
+		foreach ($results as $result)
```
Comment (Anu1601CS) [inline administrator/components/com_templates/Model/TemplateModel.php]: This loop

**Comment by astridx** [comment]
> Can we do this without creating a new loop ? Mean, here

@Anu1601CS I have put the code in the already existing loop.

@laoneo Is the handling of the date now like it should be done in Joomla?

File: administrator/components/com_templates/Model/TemplateModel.php [hunk 4]
```diff
@@ -197,8 +197,24 @@ public function getUpdatedList($state = false, $all = false, $cleanup = false)
...
 			$path = Path::clean($client->path . '/templates/' . $pk->template . base64_decode($pk->hash_id));
 
+			$tz = new \DateTimeZone(Factory::getApplication()->get('offset'));
+
 			if (file_exists($path))
 			{
+				if ((int) $pk->created_date)
+				{
+					$created_date = new Date($pk->created_date);
+					$created_date->setTimezone($tz);
```
Comment (mbabker) [inline administrator/components/com_templates/Model/TemplateModel.php]: These `setTimezone` calls should not be necessary.  Instantiating a new `Joomla\CMS\Date\Date` object internally converts the timestamp to UTC.

File: administrator/components/com_templates/Model/TemplateModel.php [hunk 5]
```diff
@@ -197,8 +197,24 @@ public function getUpdatedList($state = false, $all = false, $cleanup = false)
...
+					$pk->created_date = $created_date->toSql(true);
```
Comment (mbabker) [inline administrator/components/com_templates/Model/TemplateModel.php]: `toSql(true)` means you're getting the timestamp in the correct SQL format but localized, so this is converting from UTC to whatever timezone you're setting above.

Look at the `check` methods of various `Joomla\CMS\Table\Table` subclasses for how created and modified dates are handled.

Comment (astridx) [reply administrator/components/com_templates/Model/TemplateModel.php]: @mbabker Thank you for your help. But if I do not set the timezone, the date will be displayed in UTC.

Comment (mbabker) [reply administrator/components/com_templates/Model/TemplateModel.php]: Sorry, missed this wasn't a save operation.

The timezone shouldn't be normalized in models.  This should be done in views/layouts.  So you should use the `JHtml::date()` helper (as is done elsewhere in core) for this, as that will adjust the timezone for the user's configuration.

**Comment by astridx** [comment]
I moved normalizing of the timezone to the view

**Comment by Anu1601CS** [comment]
@dgrammatiko Javascript test fails with. 

ReferenceError: customElements is not defined
at http://localhost:9876/base/media/system/webcomponents/js/joomla-field-switcher-es5.js?8673eb07a3064c4b3028e5260638e1e20334404b:265
Firefox 59.0.0 (Ubuntu 0.0.0) ERROR
  ReferenceError: customElements is not defined
  at /drone/src/github.com/joomla-projects/gsoc18_override_management/media/system/webcomponents/js/joomla-field-switcher-es5.js:265

File: administrator/components/com_templates/tmpl/template/default_updated_files.php [hunk 6]
```diff
@@ -57,13 +57,15 @@
 									<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_templates&view=template&id=' . (int) $value->extension_id . '&file=' . $value->hash_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?>"><?php echo base64_decode($value->hash_id); ?></a>
 								</td>
 								<td>
-									<?php echo $value->created_date; ?>
+									<?php $created_date = $value->created_date; ?>
+									<?php echo $created_date > 0 ? HTMLHelper::_('date', $created_date, 'Y-m-d H:i:s') : '-'; ?>
```
Comment (laoneo) [inline administrator/components/com_templates/tmpl/template/default_updated_files.php]: There are language constants for date formats. I'm not sure but the custom field is using them.

**Comment by astridx** [comment]
Now I use language constants for date format.

**Comment by dgrammatiko** [comment]
@Anu1601CS the js tests should be ok for now

**Comment by laoneo** [comment]
Thanks.

**Comment by Anu1601CS** [comment]
@astridx Thank you very much  :)

**Comment by astridx** [comment]
Thanks. And now I understand why we store date in UTC. Up to know this was not clear to me. (I even found this unnecessary work to store a value in UTC just to convert the value back to the time zone later).
But now I had a closer look and I understand, that we do this to show the value in the user time zone - if this is set. If the user time zone is not set, we show it in the server time zone. And 
1. it's easier to do this from the fixed time-zone UTC than always need to count between different time zones. 
2. in the case of stored data you do not know in which time zone the date is stored - because it could be, that someone changed the timezone in the Joomla config. 

I hope I understood this right.

@laoneo Is everything in this repo correct again? I mean, are all issues you see here are now OK: https://github.com/joomla-projects/gsoc18_override_management/pull/52#issuecomment-408002797

**Comment by laoneo** [comment]
Yes, that's the way to go with dates, always UTC.

**Pull request merged**